### PR TITLE
Adding CancellationToken to the pipeline

### PR DESCRIPTION
**Pull request by DavidBoike**
Includes the pipeline expression compilation, and upgrades FastExpressionCompiler to 2.0 otherwise compilation does not work.

### CODE REVIEW

File: src/NServiceBus.Core/Pipeline/Behavior.cs [hunk 1]
```diff
@@ -14,18 +15,20 @@ public abstract class Behavior<TContext> : IBehavior<TContext, TContext> where T
...
         /// <param name="context">The current context.</param>
         /// <param name="next">The next <see cref="IBehavior{TContext, TContext}" /> in the chain to execute.</param>
-        public Task Invoke(TContext context, Func<TContext, Task> next)
+        /// <param name="token">A <see cref="CancellationToken"/> to observe while invoking.</param>
+        public Task Invoke(TContext context, Func<TContext, CancellationToken, Task> next, CancellationToken token)
         {
             Guard.AgainstNull(nameof(context), context);
             Guard.AgainstNull(nameof(next), next);
-            return Invoke(context, () => next(context));
+            return Invoke(context, () => next(context, token), token);
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Pipeline/Behavior.cs:23]: Should we Guard for null?

File: src/NServiceBus.Core/Pipeline/ForkConnector.cs [hunk 2]
```diff
@@ -18,15 +19,16 @@ public abstract class ForkConnector<TFromContext, TForkContext> : Behavior<TFrom
...
 
         /// <inheritdoc />
-        public sealed override Task Invoke(TFromContext context, Func<Task> next)
+        public Task Invoke(TFromContext context, Func<TFromContext, CancellationToken, Task> next, CancellationToken token)
         {
             Guard.AgainstNull(nameof(context), context);
             Guard.AgainstNull(nameof(next), next);
 
-            return Invoke(context, next, ctx => ctx.InvokePipeline());
+            return Invoke(context, next, (ctx, token2) => ctx.InvokePipeline(token2), token);
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Pipeline/ForkConnector.cs:31]: Should we Guard for null?

File: src/NServiceBus.Core/Pipeline/PipelineTerminator.cs [hunk 3]
```diff
@@ -13,18 +14,20 @@ public abstract class PipelineTerminator<T> : StageConnector<T, PipelineTerminat
...
         /// <param name="context">Context object.</param>
         /// <param name="next">Ignored since there by definition is no next behavior to call.</param>
-        public sealed override Task Invoke(T context, Func<ITerminatingContext, Task> next)
+        /// <param name="token">A <see cref="CancellationToken"/> to observe while invoking.</param>
+        public sealed override Task Invoke(T context, Func<ITerminatingContext, CancellationToken, Task> next, CancellationToken token)
         {
             Guard.AgainstNull(nameof(next), next);
 
-            return Terminate(context);
+            return Terminate(context, token);
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Pipeline/PipelineTerminator.cs:30]: Guard?

File: src/NServiceBus.Core/Pipeline/StageForkConnector.cs [hunk 4]
```diff
@@ -15,12 +16,12 @@ public abstract class StageForkConnector<TFromContext, TToContext, TForkContext>
...
     {
         /// <inheritdoc />
-        public Task Invoke(TFromContext context, Func<TToContext, Task> next)
+        public Task Invoke(TFromContext context, Func<TToContext, CancellationToken, Task> next, CancellationToken token)
         {
             Guard.AgainstNull(nameof(context), context);
             Guard.AgainstNull(nameof(next), next);
 
-            return Invoke(context, next, ctx => ctx.InvokePipeline());
+            return Invoke(context, ctx => next(ctx, token), ctx => ctx.InvokePipeline(token), token);
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Pipeline/StageForkConnector.cs:24]: Guard?

**Comment by DavidBoike** [comment]
@andreasohlund `CancellationToken` is a struct. It can't be checked for null.

**Review by andreasohlund — APPROVED** [review]

File: src/NServiceBus.Core.Tests/Pipeline/EnsureNativeBehaviors.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+                {
+                    if (baseType.IsGenericType)
+                    {
+                        var genericType = baseType.GetGenericTypeDefinition();
+                        Assert.IsFalse(genericType == typeof(Behavior<>), $"For performance reasons, built-in behavior `{type}` is not allowed to inherit from abstract class Behavior<T>. Implement IBehavior<Tin, TOut> directly, using the same context type for both TIn and TOut.");
+                    }
+                }
+            }
+        }
+    }
```
Comment (adamralph) [inline src/NServiceBus.Core.Tests/Pipeline/EnsureNativeBehaviors.cs:29]: Another possible approach, based on what I did in https://github.com/Particular/NServiceBus/pull/5850:

```suggestion    using System;
    using System.Linq;
    using NServiceBus.Pipeline;
    using NUnit.Framework;

    [TestFixture]
    public static class EnsureNativeBehvaiors
    {
        [Test]
        public static void CoreBehaviorsMustNotUseAbstractClass()
        {
            var violators = typeof(IBehavior).Assembly.GetTypes()
                .Where(UsesAbstractClass)
                .ToList();

            Console.Error.WriteLine($"Violators of {nameof(CoreBehaviorsMustNotUseAbstractClass)}:{Environment.NewLine}{string.Join(Environment.NewLine, violators)}");

            Assert.IsEmpty(violators, $"For performance reasons, built-in behaviors are not allowed to inherit from abstract class Behavior<T>. Implement IBehavior<Tin, TOut> directly, using the same context type for both TIn and TOut.");
        }

        static bool UsesAbstractClass(this Type type) =>
            type != null && type != typeof(Behavior<>) && ((type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Behavior<>)) || type.BaseType.UsesAbstractClass());
    }
```

**Review by adamralph — COMMENTED** [review]
I haven't gotten into the meat of the PR yet, i.e. `PipelineTests.cs` and `PipelineExecutionExtensions.cs`.

As a first pass, I went through all the surrounding stuff and marked it as "viewed" to get it out of the way before tackling the tough stuff, and left a few comments about minor stuff along the way.

One thing I did notice is that some token arguments are not propagated to all task returning method calls. E.g.

- https://github.com/Particular/NServiceBus/pull/5862/files#diff-262fb023a246f1741aeb2222e20b81d95398f69c274c6e6e0008acaabefb688eR20
- https://github.com/Particular/NServiceBus/pull/5862/files#diff-d2b7077b4bb8cc0eb1b1da72ea5e23511f3a7d6b20c3887dc5ae0e869f440ba4R49
- https://github.com/Particular/NServiceBus/pull/5862/files#diff-1c31635704e9325346306504a1999675cc03e0dfc339081f6d92926cff28b38fR17
- https://github.com/Particular/NServiceBus/pull/5862/files#diff-7aa96c7a93c50ceff2205c6f8f45b43be12b84a95cf536355a2e02264f59378bR17
- https://github.com/Particular/NServiceBus/pull/5862/files#diff-b33cf0eb090c12ff172f9c02b2eebeb559f62b5e496874bdaa5af9a5d3895013R25
- https://github.com/Particular/NServiceBus/pull/5862/files#diff-8159993de3368bd418156794663139a1808874145aed57c59e40aec719a9fd60R30
- https://github.com/Particular/NServiceBus/pull/5862/files#diff-8159993de3368bd418156794663139a1808874145aed57c59e40aec719a9fd60R52
- https://github.com/Particular/NServiceBus/pull/5862/files#diff-37f32e8c08299d2e35351cdcfb26a5f105b7c19d8053fe801c781ed3c2eec223R79

Is that intentional?

BTW - regarding the parameter naming, my initial instinct was to use the terse `token`, but it seems that the more common convention (e.g. used in the BCL) is `cancellationToken` and I wrote a test to enforce that in https://github.com/Particular/NServiceBus/pull/5850. I'm a bit sad 🐼 at the longer name, but it may be better to go with that convention.

File: src/NServiceBus.Core.Tests/Pipeline/PipelineTests.cs [hunk 6]
```diff
@@ -56,25 +103,58 @@ public async Task ShouldNotCacheContext()
...
             context.Extensions.Set(ExtendableExtensions.RunSpecificKey, 2);
 
-            await pipeline.Invoke(context);
+            await pipeline.Invoke(context, default);
 
             Approver.Verify(stringWriter.ToString());
         }
 
+        [Test]
+        public async Task ShouldNotCacheCancellationToken()
```
Comment (adamralph) [inline src/NServiceBus.Core.Tests/Pipeline/PipelineTests.cs:126]: Why is are both this and `ShouldNotCacheTokenBetweenInvocations` required?

File: src/NServiceBus.Core.Tests/Pipeline/PipelineTests.cs [hunk 7]
```diff
@@ -56,25 +103,58 @@ public async Task ShouldNotCacheContext()
...
+            Assert.ThrowsAsync<OperationCanceledException>(() => pipeline.Invoke(context, new CancellationToken(true)));
+
+            stringWriter.WriteLine("Run 2");
+
+            context = new TestableTransportReceiveContext();
+            context.Extensions.Set<IPipelineCache>(new FakePipelineCache());
+            context.Extensions.Set(ExtendableExtensions.RunSpecificKey, 2);
+
+            // Should not throw
+            await pipeline.Invoke(context, CancellationToken.None);
```
Comment (adamralph) [inline src/NServiceBus.Core.Tests/Pipeline/PipelineTests.cs:155]: Should we be explicit here and pass `new CancellationToken(false)` (as in `ShouldNotCacheTokenBetweenInvocations`)?

File: src/NServiceBus.Core/Pipeline/PipelineExecutionExtensions.cs [hunk 8]
```diff
@@ -48,30 +50,31 @@ public static Delegate CreatePipelineExecutionExpression(this IBehavior[] behavi
...
-                var inContextParameter = ExpressionInfo.Parameter(inContextType, $"context{i}");
+                var inContextParameter = Parameter(inContextType, $"context{i}");
+                var cancellationTokenParameter = Parameter(typeof(CancellationToken), $"token{i}");
 
                 if (i == behaviorCount)
                 {
-                    var outContextType = genericArguments[1];
-                    var doneDelegate = CreateDoneDelegate(outContextType, i);
-                    lambdaExpression = CreateBehaviorCallDelegate(currentBehavior, methodInfo, inContextParameter, doneDelegate, expressions);
+                    inContextType = genericArguments[1];
```
Comment (adamralph) [inline src/NServiceBus.Core/Pipeline/PipelineExecutionExtensions.cs:58]: You'll have to walk me through the changes in this class, starting here. 🙂

**Comment by DavidBoike** [comment]
After change in direction, closing this PR. One thing (ensuring internal behaviors directly use IBehavior) was not technically part of making the pipeline token-aware—that has been extracted to https://github.com/Particular/NServiceBus/pull/5882.

**Pull request closed without merging**